src/util/expr_iterator.h

```diff
@@ -0,0 +1,247 @@
+/// class. It's only purpose is to provide common functions for other
+/// DFS iterators
+/// Do override .push_expr method of this class, but when doing so
+/// make this class a friend so it has acess to that overriden .push_expr
+/// method in the child class
+template<typename depth_iterator_t>
+class depth_iterator_baset:
+public std::iterator<std::forward_iterator_tag, const exprt, void>
+{
+public:
+  bool operator==(const depth_iterator_t &other) const
+  {
+    return m_stack.size()==other.m_stack.size() &&
+           std::equal(m_stack.begin(),
+                      m_stack.end(),
+                      other.m_stack.begin());
+  }
+
+  bool operator!=(const depth_iterator_t &other) const
+  {
+    return !(*this==other);
+  }
+
+  /// Preincrement operator
+  /// Do not call on the end() iterator
+  depth_iterator_t &operator++()
+  {
+    PRECONDITION(!m_stack.empty());
+    while(true)
+    {
+      if(m_stack.back().it==m_stack.back().end)
+      {
+        m_stack.pop_back();
+        if(m_stack.empty())
+        {
+          break;
+        }
+      }
+      else if(this->downcast().push_expr(*m_stack.back().it))
+      {
+        break;
+      }
+      m_stack.back().it++;
+    }
+    return this->downcast();
+  }
+
+  /// Dereferencing end() iterator is undefined behaviour
+  const exprt &operator*() const
+  {
+    PRECONDITION(!m_stack.empty());
+    return m_stack.back().expr.get();
+  }
+
+  /// Dereferencing end() iterator is undefined behaviour
+  const exprt *operator->() const
+  {
+    return &**this;
+  }
+
+protected:
+  /// Create end iterator
+  depth_iterator_baset() { }
+
+  /// Create begin iterator, starting at the supplied node
+  explicit depth_iterator_baset(const exprt &root)
+  {
+    this->downcast().push_expr(root);
+  }
+
+  /// Obtain non-const exprt reference. Breaks the tree as a side
+  /// effect. To be called only if a non-const exprt was passed into
+  /// the constructor Do not call on the end() iterator
+  exprt &mutate()
+  {
+    PRECONDITION(!m_stack.empty());
+    auto expr=std::ref(const_cast<exprt&>(m_stack.front().expr.get()));
+    for(auto& state : m_stack)
+    {
+      if(state.operands)
+      {
+        if(!(state==m_stack.back()))
+          expr=std::ref(const_cast<exprt&>(*state.it));
+      }
+      else
+      {
+        auto &operands=expr.get().operands();
+        const auto i=operands.size()-(state.end-state.it);
+        const auto it=operands.begin()+i;
+        state.expr=expr.get();
+        state.it=it;
+        state.end=operands.end();
+        state.operands=&operands;
+        if(!(state==m_stack.back()))
+        {
+          expr=std::ref(*it);
+        }
+      }
+    }
+    return expr.get();
+  }
+
+  /// Pushes expression onto the stack
+  /// If overriden, this function should be called from the inheriting
+  /// class by the override function
+  /// \return true if element was successfully pushed onto the stack,
```

Comment:
Yes, please add a comment: return true to iterate over this tree node, or false to skip it.